optimum/habana/transformers/trainer.py

```diff
@@ -739,16 +731,15 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args):
                 if step % args.gradient_accumulation_steps == 0:
                     self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
 
-                if (
-                    ((step + 1) % args.gradient_accumulation_steps != 0)
-                    and args.local_rank != -1
-                    and args._no_sync_in_gradient_accumulation
-                ):
-                    # Avoid unnecessary DDP synchronization since there will be no backward pass on this example.
-                    with model.no_sync():
-                        tr_loss_step = self.training_step(model, inputs)
-                else:
-                    tr_loss_step = self.training_step(model, inputs)
+                tr_loss_step = self._take_training_step_fwbwd(model, inputs)
```

Comment:
If I understood correctly, `_take_training_step_fwbwd` deals with HPU graphs only, so I think we should keep `training_step` for non-HPU-graphs cases.

Reply:
I concur. I will re-think how to code this elegantly.